### PR TITLE
[kitsune] Add error management when the kitsune sever fails

### PR DESCRIPTION
**Pull request by acs**
Add error management when the kitsune sever fails providing questions in a page

### CODE REVIEW

File: perceval/backends/kitsune.py [hunk 1]
```diff
@@ -107,8 +109,28 @@ def fetch(self, offset=DEFAULT_OFFSET):
...
+
+        while True:
+            try:
+                raw_questions = next(questions_page)
+            except StopIteration:
+                break
+            except requests.exceptions.HTTPError as e:
+                # Continue with the next page if it is a 500 error
+                if e.response.status_code == 500:
+                    traceback.print_exc()
```
Comment (sduenas) [inline perceval/backends/kitsune.py]: This should be printed to the log not to the error output.

File: perceval/backends/kitsune.py [hunk 2]
```diff
@@ -107,8 +109,28 @@ def fetch(self, offset=DEFAULT_OFFSET):
...
+                    logging.error("Problem getting Kitsune questions. Loosing %i questions. Going to the next page.", KitsuneClient.ITEMS_PER_PAGE)
```
Comment (sduenas) [inline perceval/backends/kitsune.py]: Use `logger` object, not `logging`. You also have to split the line. It is too long.

File: tests/test_kitsune.py [hunk 3]
```diff
@@ -211,6 +216,18 @@ def test_fetch_empty(self):
 
         self.assertEqual(len(questions), 0)
 
+    @httpretty.activate
```
Comment (sduenas) [inline tests/test_kitsune.py]: I don't understand this test. I know that it is going to fail when an `HTTPError` exception is raised unless it is a 500 error, that would continue without failing. Your tests should take into account that behaviour.

File: tests/test_kitsune.py [hunk 4]
```diff
@@ -211,6 +216,18 @@ def test_fetch_empty(self):
...
+    def test_fetch_server_error(self):
+        """Test whether it works when the server fails"""
+
+        HTTPServer.routes(empty=True)
+
+        kitsune = Kitsune(KITSUNE_SERVER_URL)
+        offset = (KITSUNE_SERVER_FAIL_PAGE-1)*KITSUNE_ITEMS_PER_PAGE
```
Comment (sduenas) [inline tests/test_kitsune.py]: Add separation between operators.

File: tests/test_kitsune.py (hunk 3 above)
Comment (acs) [inline tests/test_kitsune.py]: In this tests we access the page (69) which generates the 500 error that it is managed inside the backend, and no questions are returned. Then, the next page (70) is retrieved which does not exists and generates a 404 (not found) error which is captured by the assert. Then, the total number of pages is checked, which must be 0.

File: perceval/backends/kitsune.py (hunk 1 above)
Comment (acs) [inline perceval/backends/kitsune.py]: Fixed!

File: tests/test_kitsune.py (hunk 4 above)
Comment (acs) [inline tests/test_kitsune.py]: Fixed!

File: perceval/backends/kitsune.py [hunk 5]
```diff
@@ -134,11 +157,13 @@ def fetch(self, offset=DEFAULT_OFFSET):
                 nquestions += 1
                 self._push_cache_queue('{}') # Mark with empty dict end of question
 
-            logger.debug("Questions: %i/%i", nquestions, tquestions)
+            logger.debug("Questions: %i/%i", nquestions+offset, tquestions)
```
Comment (sduenas) [inline perceval/backends/kitsune.py]: Put spaces between operators.

File: tests/test_kitsune.py [hunk 6]
```diff
@@ -87,6 +89,12 @@ def request_callback(method, uri, headers):
                     body = mozilla_questions_2
                 else:
                     body = mozilla_question_answers_1
+            elif page == str(KITSUNE_SERVER_FAIL_PAGE):
+                # To tests for Internal Server Error
+                return (500, headers, '')
+            elif page == str(KITSUNE_SERVER_FAIL_PAGE+1):
```
Comment (sduenas) [inline tests/test_kitsune.py]: Spaces between operators.

File: tests/test_kitsune.py [hunk 7]
```diff
@@ -211,6 +219,18 @@ def test_fetch_empty(self):
...
         self.assertEqual(len(questions), 0)
 
+    @httpretty.activate
+    def test_fetch_server_error(self):
+        """Test whether it works when the server fails"""
+
+        HTTPServer.routes(empty=True)
+
+        kitsune = Kitsune(KITSUNE_SERVER_URL)
+        offset = (KITSUNE_SERVER_FAIL_PAGE-1) * KITSUNE_ITEMS_PER_PAGE
```
Comment (sduenas) [inline tests/test_kitsune.py]: Same comment as above here. Put spaces between operators.

File: tests/test_kitsune.py (hunk 3 above)
Comment (sduenas) [inline tests/test_kitsune.py]: This test is much better. The server will fail but the backend will continue fetching items.

File: perceval/backends/kitsune.py (hunk 5 above)
Comment (acs) [inline perceval/backends/kitsune.py]: Fixed! Ups!

File: tests/test_kitsune.py (hunk 3 above)
Comment (acs) [inline tests/test_kitsune.py]: Great!

**Comment by sduenas** [comment]
Merged!

**Pull request closed without merging**